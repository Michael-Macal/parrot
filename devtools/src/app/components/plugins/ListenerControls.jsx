import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TextField from 'material-ui/TextField';
import RaisedButton from 'material-ui/RaisedButton';

import fetchApi from '../../utils/fetchApi';

export default class ListenerControls extends Component {
  static propTypes = {
    url: PropTypes.string.isRequired,
  };

  state = {
    scenarioName: '',
    error: false,
    isListening: false,
  };

  // Check if listener is currently running
  componentDidMount() {
    fetchApi(this.props.url, '/parrot/listen')
      .then(resp => resp.json())
      .then(config => this.setState(config))
      .catch(error => this.setState({ error }));
  }

  setName = (event, scenarioName) => this.setState({ scenarioName });

  toggleListening = () => {
    const { isListening, scenarioName } = this.state;
    fetchApi(this.props.url, '/parrot/listen', {
      method: 'PUT',
      headers: {
        Accept: 'application/json, text/plain, */*',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        action: isListening ? 'STOP' : 'START',
        scenarioName,
      }),
    })
      .then(() => this.setState({ isListening: !isListening }))
      .catch(error => this.setState({ error }));
  };

  render() {
    const { isListening, scenarioName, error } = this.state;
    const toggleLabel = `${isListening ? 'Stop' : 'Start'} Listening`;
    return (
      <div>
        {error && <div><pre>{error}</pre></div>}
        <TextField
          name="Scenario Name"
          floatingLabelText="Scenario Name"
          value={scenarioName}
          onChange={this.setName}
          disabled={isListening}
        />
        <RaisedButton
          label={toggleLabel}
          primary={!isListening}
          secondary={isListening}
          onTouchTap={this.toggleListening}
          disabled={scenarioName === ''}
        />
      </div>
    );
  }
}
