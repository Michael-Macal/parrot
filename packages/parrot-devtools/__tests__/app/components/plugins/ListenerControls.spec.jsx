import React from 'react';
import { shallow } from 'enzyme';

import ListenerControls from '../../../../src/app/components/plugins/ListenerControls';

jest.mock('../../../../src/app/utils/fetchApi');

describe('ListenerControls component', () => {
  it('renders', () => {
    const props = {
      url: 'http://localhost:8080',
    };
    const wrapper = shallow(<ListenerControls {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
