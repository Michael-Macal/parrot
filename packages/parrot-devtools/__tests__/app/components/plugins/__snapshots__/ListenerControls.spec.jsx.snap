// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListenerControls component renders 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <TextField
        disabled={false}
        floatingLabelFixed={false}
        floatingLabelText="Scenario Name"
        fullWidth={false}
        multiLine={false}
        name="Scenario Name"
        onChange={[Function]}
        rows={1}
        type="text"
        underlineShow={true}
        value=""
    />
    <RaisedButton
        disabled={true}
        fullWidth={false}
        label="Start Listening"
        labelPosition="after"
        onTouchTap={[Function]}
        primary={true}
        secondary={false}
    />
</div>,
  "nodes": Array [
    <div>
      <TextField
            disabled={false}
            floatingLabelFixed={false}
            floatingLabelText="Scenario Name"
            fullWidth={false}
            multiLine={false}
            name="Scenario Name"
            onChange={[Function]}
            rows={1}
            type="text"
            underlineShow={true}
            value=""
      />
      <RaisedButton
            disabled={true}
            fullWidth={false}
            label="Start Listening"
            labelPosition="after"
            onTouchTap={[Function]}
            primary={true}
            secondary={false}
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ListenerControls
        url="http://localhost:8080"
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ListenerControls {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "url": "http://localhost:8080",
        },
        "refs": Object {},
        "setName": [Function],
        "state": Object {
          "error": false,
          "isListening": false,
          "scenarioName": "",
        },
        "toggleListening": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <TextField
                    disabled={false}
                    floatingLabelFixed={false}
                    floatingLabelText="Scenario Name"
                    fullWidth={false}
                    multiLine={false}
                    name="Scenario Name"
                    onChange={[Function]}
                    rows={1}
                    type="text"
                    underlineShow={true}
                    value=""
          />
          <RaisedButton
                    disabled={true}
                    fullWidth={false}
                    label="Start Listening"
                    labelPosition="after"
                    onTouchTap={[Function]}
                    primary={true}
                    secondary={false}
          />
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <TextField
                    disabled={false}
                    floatingLabelFixed={false}
                    floatingLabelText="Scenario Name"
                    fullWidth={false}
                    multiLine={false}
                    name="Scenario Name"
                    onChange={[Function]}
                    rows={1}
                    type="text"
                    underlineShow={true}
                    value=""
          />
          <RaisedButton
                    disabled={true}
                    fullWidth={false}
                    label="Start Listening"
                    labelPosition="after"
                    onTouchTap={[Function]}
                    primary={true}
                    secondary={false}
          />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ListenerControls
    url="http://localhost:8080"
/>,
}
`;
